File: php/01_factory/main.php
<?php

define("PaymentGatewayTypeStripe", 1);
define("PaymentGatewayTypePaypal", 2);

interface PaymentGateway {
    public function ProcessPayment();
}

class StripeGateway implements PaymentGateway {
    public function ProcessPayment() {
        echo "\nStripe Gateway : Process Payment\n";
    }
}

class PaypalGateway implements PaymentGateway {
    public function ProcessPayment() {
        echo "\nPaypal Gateway : Process Payment\n";
    }
}

class Factory {
    public function factoryMethod($gatewayType) {
        switch($gatewayType) {
            case 1:
                return new StripeGateway();
                break;
            case 2:
                return new PaypalGateway();
                break;
        }
    }
}

$factory = new Factory();

$obj = $factory->factoryMethod(PaymentGatewayTypeStripe);
$obj->ProcessPayment();

$obj = $factory->factoryMethod(PaymentGatewayTypePaypal);
$obj->ProcessPayment();

?>

File: go/01_factory/main.go
package main

import "fmt"

type PaymentGatewayType int

const (
	GatewayTypeStripe PaymentGatewayType = 1
	GatewayTypePaypal PaymentGatewayType = 2
)

type PaymentGateway interface {
	ProcessPayment(amount float64)
}

type StripeGateway struct{}

func (*StripeGateway) ProcessPayment(amount float64) {
	fmt.Println("Processing payment using Stripe Gateway : ", amount)
}

type PaypalGateway struct{}

func (*PaypalGateway) ProcessPayment(amount float64) {
	fmt.Println("Processing payment using PayPal Gateway : ", amount)
}

// factoryMethod - instantiate a class bsaed on the passed parameter
func FactoryMethod(gatewayType PaymentGatewayType) PaymentGateway {
	switch gatewayType {
	case GatewayTypeStripe:
		return &StripeGateway{}

	case GatewayTypePaypal:
		return &PaypalGateway{}
	}

	return nil
}

func main() {
	// object of stripe is created
	paymentObj := FactoryMethod(GatewayTypeStripe)
	paymentObj.ProcessPayment(10.10)

	// object of payPal is created
	paymentObj = FactoryMethod(GatewayTypePaypal)
	paymentObj.ProcessPayment(10.20)
}
